extension/basicauth/basicauth.go

```diff
@@ -0,0 +1,140 @@
+	"go.opentelemetry.io/collector/config/configauth"
+)
+
+var _ configauth.ServerAuthenticator = (*BasicAuth)(nil)
+
+type BasicAuth struct {
+	htpasswd  HtpasswdSettings
+	matchFunc matchFunc
+}
+
+type matchFunc func(username, password string) bool
+
+func (ba *BasicAuth) Start(ctx context.Context, host component.Host) error {
+	inlineHtp, err := htpasswd.NewFromReader(strings.NewReader(ba.htpasswd.Inline), htpasswd.DefaultSystems, nil)
+	if err != nil {
+		return fmt.Errorf("read htpasswd from inline: %w", err)
+	}
+	ba.matchFunc = inlineHtp.Match
+
+	if ba.htpasswd.File != "" {
+		fileHtp, err := htpasswd.New(ba.htpasswd.File, htpasswd.DefaultSystems, nil)
+		if err != nil {
+			return fmt.Errorf("read htpasswd from file: %w", err)
+		}
+
+		matchFunc := ba.matchFunc
+		ba.matchFunc = func(username, password string) bool {
+			// Inline takes precedence over the file.
+			return matchFunc(username, password) || fileHtp.Match(username, password)
+		}
+	}
+
+	return nil
+}
+
+func (ba *BasicAuth) Shutdown(ctx context.Context) error {
+	return nil
+}
+
+var (
+	errNoAuth              = errors.New("no basic auth provided")
+	errInvalidCredentials  = errors.New("invalid credentials")
+	errInvalidSchemePrefix = errors.New("invalid authorization scheme prefix")
+	errInvalidFormat       = errors.New("invalid authorization format")
+)
+
+func (ba *BasicAuth) Authenticate(ctx context.Context, headers map[string][]string) (context.Context, error) {
+	authHeaders := headers["authorization"]
+	if len(authHeaders) == 0 {
+		return ctx, errNoAuth
+	}
+
+	authData, err := parseBasicAuth(authHeaders[0])
+	if err != nil {
+		return ctx, err
+	}
+
+	if !ba.matchFunc(authData.username, authData.password) {
+		return ctx, errInvalidCredentials
+	}
+
+	cl := client.FromContext(ctx)
+	cl.Auth = authData
+	return client.NewContext(ctx, cl), nil
+}
+
+// See: https://github.com/golang/go/blob/1a8b4e05b1ff7a52c6d40fad73bcad612168d094/src/net/http/request.go#L950
+func parseBasicAuth(auth string) (*authData, error) {
+	const prefix = "Basic "
+	if len(auth) < len(prefix) || !strings.EqualFold(auth[:len(prefix)], prefix) {
+		return nil, errInvalidSchemePrefix
+	}
+
+	encoded := auth[len(prefix):]
+	decodedBytes, err := base64.StdEncoding.DecodeString(encoded)
+	if err != nil {
+		return nil, errInvalidFormat
+	}
+	decoded := string(decodedBytes)
+
+	si := strings.IndexByte(decoded, ':')
+	if si < 0 {
+		return nil, errInvalidFormat
+	}
+
+	return &authData{
+		username: decoded[:si],
+		password: decoded[si+1:],
+		raw:      encoded,
+	}, nil
+}
+
+var _ client.AuthData = (*authData)(nil)
+
+type authData struct {
+	username string
+	password string
+	raw      string
+}
+
+func (a *authData) GetAttribute(name string) interface{} {
+	switch name {
+	case "subject":
+		return a.username
+	case "raw":
+		return a.raw
+	default:
+		return nil
+	}
+}
+
+func (*authData) GetAttributeNames() []string {
+	return []string{"subject", "raw"}
```

Comment:
I think `username` would be more appropriate here instead of `subject`, no? What does httpd use?

Reply:
I thought there is some convention in naming attributes so there is an interoperability between authenticators. I can make it `username` or just make it support both `subject` and `username`.

Reply:
There are different auth standards with slightly different semantics for the names. This is the reason why we have a map and provide the attribute names instead of using a fixed set of attributes and names. So, use whatever is commonly used when configuring htpasswd elsewhere.

Reply:
Made it just `username`.